r2-core/src/main/java/com/linkedin/r2/transport/http/client/EvictingCircularBuffer.java

```diff
@@ -0,0 +1,245 @@
+ */
+@SuppressWarnings("serial")
+public class EvictingCircularBuffer implements CallbackStore
+{
+  private Duration _ttl;
+  private final ArrayList<Callback<None>> _callbacks = new ArrayList<>();
+  private final ArrayList<Instant> _ttlBuffer = new ArrayList<>();
+  private final ArrayList<ReentrantReadWriteLock> _elementLocks = new ArrayList<>();
+  private final AtomicInteger _readerPosition = new AtomicInteger();
+  private final AtomicInteger _writerPosition = new AtomicInteger();
+  private final Clock _clock;
+
+
+  public EvictingCircularBuffer(int capacity, int ttl, ChronoUnit ttlUnit)
+  {
+    this(capacity, ttl, ttlUnit, SystemClock.instance());
+  }
+
+  EvictingCircularBuffer(int capacity, int ttl, ChronoUnit ttlUnit, Clock clock)
+  {
+    setCapacity(capacity);
+    setTtl(ttl, ttlUnit);
+    _clock = clock;
+  }
+
+  /**
+   * Adds the supplied Callback to the internal circular buffer. If the buffer is full, the oldest Callback in the buffer
+   * will be overwritten. Calls to put always succeed, and there is no guarantee that Callbacks submitted through the put()
+   * method will be subsequently returned by the get() method.
+   *
+   * @param toAdd Callback that is to be possibly returned by later calls to get()
+   */
+  public void put(Callback<None> toAdd)
+  {
+    int writerPosition = getAndBumpWriterPosition();
+    ReentrantReadWriteLock thisLock = _elementLocks.get(writerPosition);
+    thisLock.writeLock().lock();
+    try
+    {
+      _callbacks.set(writerPosition, toAdd);
+      _ttlBuffer.set(writerPosition, Instant.ofEpochMilli(_clock.currentTimeMillis()));
+    } finally {
+      thisLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Returns a Callback previously stored in the circular buffer through the put() method. Callbacks are generally returned
+   * in the order they were received, but in cases of get() throughput in excess of put() throughput, previously returned
+   * Callbacks will be sent. Oldest age of returned Callbacks is configurable through the ttl param in the constructor.
+   *
+   * Calls to get() will always succeed as long as calls to put() continue at a cadence within the ttl duration. When write
+   * throughput has dropped to zero, get() will eventually throw NoSuchElementException once the circular buffer has become
+   * fully pruned through expired ttl.
+   * @return Callback
+   * @throws NoSuchElementException if internal circular buffer is empty
+   */
+  public Callback<None> get()
+  {
+    int thisReaderPosition = getAndBumpReaderPosition();
+    ReentrantReadWriteLock thisLock = _elementLocks.get(thisReaderPosition);
+
+    thisLock.readLock().lock();
+    Callback<None> callback;
+    Instant ttl;
+    try {
+      callback = _callbacks.get(thisReaderPosition);
+      ttl = _ttlBuffer.get(thisReaderPosition);
+    }
+    finally
+    {
+      thisLock.readLock().unlock();
+    }
+
+    if (callback == null)
+    {
+      if (thisReaderPosition == 0)
```

Comment:
Reaching the position 0 doesn't necessarily mean that the buffer is empty. It could also be the case that the element at position 0 has expired (but the element at position 1 is not necessarily expired). What we can do is probably to add an atomic counter of the number of items in the buffer. When put, increment the counter; when expire, decrement the counter.

Reply:
Chatted about this on Zoom. Opted to implement a much cleaner solution where we iterate over a while loop that increments an int until it reaches the size of the buffer.